File: singly_linked_list/python/ds.py
"""SinglyLinkedList
Running time for get operations is O(N)
Running time for append, prepend is O(1)
Running time for remove, remove_at is effectively O(N) because the get traversal happens
"""

from typing import Generic, Optional, TypeVar

T = TypeVar("T")


class Node(Generic[T]):
    _value: T
    _next: "Optional[Node[T]]"

    def __init__(self, value: T, next: "Optional[Node[T]]" = None) -> None:
        self._value = value
        self._next = next

    @property
    def value(self) -> T:
        return self._value

    @value.setter
    def value(self, v: T) -> None:
        self._value = v

    @property
    def next(self) -> "Optional[Node[T]]":
        return self._next

    @next.setter
    def next(self, v: "Node[T]") -> None:
        self._next = v


class SinglyLinkedList(Generic[T]):
    """Stores a reference to the head node
    and provides the methods needed to work with the list.
    """

    def __init__(self) -> None:
        self._head: Optional[Node[T]] = None

    def __str__(self) -> str:
        vals = []
        node = self.head
        while node:
            vals.append(node.value)
            node = node.next
        return "->".join([str(v) for v in vals])

    @property
    def head(self) -> Optional[Node[T]]:
        return self._head

    @head.setter
    def head(self, value: Optional[Node[T]]) -> None:
        self._head = value

    @property
    def tail(self) -> Optional[Node[T]]:
        if not self.head:
            return None
        else:
            node = self.head
            while node.next:
                node = node.next
            return node

    def length(self) -> int:
        """Return the length of the list
        Args:
            None

        Returns:
            The integer length of the list
        """
        node = self.head
        count = 0
        while node:
            node = node.next
            count += 1
        return count

    def get(self, idx: int) -> T:
        """Get the value of the node at the specified index.

        Args:
            idx (int) The index of the node you're seeking a value for
        Returns:
            The value of the node at the index
        Raises
            IndexError if attempting to access an index that doesn't exist
        """

        node = self.head
        i = 0
        while i < idx:
            if not node:
                msg = f"Node at index {i} not found when seeking node at index {idx}"
                raise IndexError(msg)

            if node.next:
                node = node.next
                i += 1

        if not node:
            msg = f"Node at index {i} not found when seeking node at index {idx}"
            raise IndexError(msg)
        return node.value

    def remove(self, value: T) -> T:
        """Remove the first node which has the provided value
        Args:
            value (T) The value you're searching for
        Returns:
            The value of the node you found. Should be the same you passed in
        Raises:
            ValueError if no node is found with the provided value
        """
        parent_node = None
        node = self.head
        while node:
            if value == node.value:
                if not parent_node and node:
                    out = node.value
                    self.head = node.next
                    return out
                elif parent_node and node:
                    out = node.value
                    parent_node.next = node.next  # type: ignore None check in elif
                    return out
                else:
                    msg = f"Unreachable: we can never have a parent_node without child"
                    raise NotImplementedError(msg)
            else:
                parent_node = node
                node = node.next
        msg = f"The value {value} was not found in the list"
        raise ValueError(msg)

    def remove_at(self, idx: int) -> T:
        """Remove the node at the specified index
        Args:
            idx (int) The index of the node you'd like to remove

        Returns:
            The value of the node at the index you just removed

        Raises:
            IndexError if the index doesn't exist
        """
        parent_node = None
        node = self.head
        i = 0
        while i < idx:
            if not node:
                msg = f"Node at index {i} not found when seeking node at index {idx}"
                raise IndexError(msg)

            if node.next:
                parent_node = node
                node = node.next
                i += 1

        if node and not parent_node:
            assert i == 0
            out = self.head.value  # type: ignore - the None check is in the if
            self.head = node.next
            return out
        elif node and parent_node:
            next_node = node.next
            parent_node.next = next_node  # type: ignore - the None check is in the elif
            out = node.value
            return out
        else:
            msg = f"Unreachable: we can never have a parent_node without child"
            raise NotImplementedError(msg)

    def append(self, value: T) -> None:
        """Add a new node to the end of the list

        Args:
            value (T) The value you want to append
        Returns:
            None
        Raises:
            AssertionError if the existing last node already has a next node
        """

        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return None
        else:
            assert self.tail != None
            assert self.tail.next == None
            self.tail.next = new_node

    def prepend(self, value: T) -> None:
        """Add a new node to the beginning of the list
        Args:
            value (T) The value to prepend
        Returns:
            None
        """
        new_node = Node(value)
        if self.head:
            new_node.next = self.head
        self.head = new_node

File: singly_linked_list/lua/ds.lua
--- Construct a new node for the provided value
-- @param val The value to store in the node
-- @return A table representing the new node with keys (value, next)
local function node(val)
  return { ["value"] = val, ["next"] = nil }
end

--- Add a new node to the front of the list
-- This as running time of O(1)
-- @param head The head node of the linked list
-- @param v The value to add to a node at the front of the list
-- @return The new head node
local function prepend(head, v)
  local new_head = node(v)
  if head then
    new_head.next = head
  end
  return new_head
end

--- Get the last node in the list
-- This as running time of O(N) as it has to traverse to the end
-- @param head The head node of the linked list
-- @return The tail node or nil
local function tail(head)
  local n = head
  while n do
    if n.next then
      n = n.next
    else
      break
    end
  end
  return n
end

--- Add a new node to the end of the list
-- This as running time of O(N)
-- @param head The head node of the linked list
-- @param v The value to add to a node at the end of the list
-- @return The same head node
local function append(head, v)
  local new_node = node(v)
  local n = tail(head)
  if n then
    n.next = new_node
  else
    head = new_node
  end
  return head
end

--- Get a last node in the list at the provided index
-- This as running time of O(N)
-- @param head The head node of the linked list
-- @param idx The index you're seeking the value at
-- @return The head node
-- @return The value of the node at the provided index
local function get(head, idx)
  local n = head
  local i = 0
  while i < idx do
    if n.next then
      n = n.next
      i = i + 1
    else
      return nil, nil
    end
  end
  return head, n.value
end

--- Remove the node at the provided index
-- This as running time of O(N)
-- @param head The head node of the linked list
-- @param idx The index of the node you're removing
-- @return The (potentially new) head node
-- @return The value at the idx you just removed
local function remove_at(head, idx)
  local p = nil
  local n = head
  local i = 0
  while i < idx do
    if n.next then
      p = n
      n = n.next
      i = i + 1
    else
      return nil
    end
  end

  if p then
    -- i > 0
    if n then
      if n.next then
        p.next = n.next
        return head, n.value
      else
        p.next = nil
        return head, n.value
      end
    else
      return nil, nil
    end
  else
    -- i = 0
    if n then
      if n.next then
        return n.next, n.value
      else
        return nil, n.value
      end
    else
      return nil, nil
    end
  end
end

--- Get the number of nodes in the list
-- This as running time of O(N)
-- @param head The head node of the linked list
-- @return The number of nodes in the list
local function length(head)
  local n = head
  local count = 0
  while n do
    count = count + 1
    if n.next then
      n = n.next
    else
      break
    end
  end
  return count
end

--- Remove the first node with the provided value
-- This as running time of O(N)
-- @param head The head node of the linked list
-- @param val The value of the node you're removing
-- @return The (potentially new) head node
-- @return The value at the node you just removed
local function remove(head, val)
  local p = nil
  local n = head
  while n do
    if n.value == val then
      if p then
        if n then
          if n.next then
            p.next = n.next
            return head, val
          else
            return head, nil
          end
        else
        end
      else
        -- i = 0
        if n then
          if n.next then
            return n.next, val
          else
            return nil, val
          end
        else
          return nil, nil
        end
      end
    else
      p = n
      n = n.next
    end
  end
  return head, nil
end

--- Print a string repr of the linked list
-- @param head The head node of the linked list
-- @return A string e.g. "5 -> 7"
local function repr(head)
  local out = ""
  local n = head
  while n do
    if out ~= "" then
      out = out .. "->" .. n.value
    else
      out = out .. n.value
    end
    n = n.next
  end
  return out
end

-- @class LinearSearchModule
local M = {}
M.prepend = prepend
M.append = append
M.get = get
M.remove_at = remove_at
M.length = length
M.remove = remove
M.repr = repr
return M
